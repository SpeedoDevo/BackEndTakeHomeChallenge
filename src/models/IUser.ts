import IUserModel from './IUserModel'

export default interface IUser {
  getId(): number
  getName(): string
  getUsername(): string
  getAvatar(): string | undefined
  updateName(name: string): this
  updateAvatar(avatar?: string): this
  toObject(): IUserModel
}
