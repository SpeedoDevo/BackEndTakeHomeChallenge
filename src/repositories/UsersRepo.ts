import IUser from '../models/IUser'
import User from '../models/User'

const users = [
  {
    id: 1,
    name: 'Daenerys Targaryen',
    username: 'thequeenbee',
  },
  {
    id: 2,
    name: 'Jon Snow',
    username: 'forthebrotherhood'
  },
  {
    id: 3,
    name: 'Sansa Stark',
    username: 'joffery.fan'
  },
  {
    id: 4,
    name: 'Ned Stark',
    username: 'honouraboveall'
  },
  {
    id: 5,
    name: 'Cersei Lannister',
    username: 'bestmuminwesteros'
  }
]

export const getLastIdUsers = () => users.reduce((prev, current) => (prev.id > current.id) ? prev : current).id
export const getAllUsers = () => users.map(user => new User(user))
export const findUserById = (id: number) => users.find(user => user.id === id)
export const addUser = (user: IUser) => users.push({
  id: user.getId(),
  name: user.getName(),
  username: user.getUsername(),
})
