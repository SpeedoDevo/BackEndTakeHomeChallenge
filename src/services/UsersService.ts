import BadRequestError from '../errors/BadRequestError'
import IUser from '../models/IUser'
import IUserModel from '../models/IUserModel'
import User from '../models/User'
import {
  addUser,
  getAllUsers,
  getLastIdUsers,
  getUserById,
  getUsersByQuery,
  updateUser,
} from '../repositories/UsersRepo'

export default class UsersService {
  /**
   * Returns all users
   */
  public getAll(): Array<IUser> {
    return getAllUsers()
  }

  public search(query: string): Array<IUser> {
    return getUsersByQuery(query)
  }

  /**
   * Creates new user
   */
  public create(name: string, username: string): IUser {
    if (name.length > 50) {
      throw new BadRequestError(
        'It looks like your name is too long. Maybe you should change it? :)',
      )
    }

    const lastId = getLastIdUsers()
    const user = new User({
      name,
      username,
      id: lastId + 1,
    })

    addUser(user)
    return user
  }

  public update(source: IUserModel): IUser {
    const user = getUserById(source.id)

    if (user) {
      user.updateName(source.name)
      user.updateAvatar(source.avatar)
      updateUser(user)
    } else {
      throw new BadRequestError()
    }

    return user
  }
}
