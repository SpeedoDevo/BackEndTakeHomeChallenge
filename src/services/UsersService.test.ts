import IUser from '../models/IUser'
import * as repo from '../repositories/UsersRepo'
import UsersService from './UsersService'

describe('UsersService', () => {
  let service: UsersService

  beforeEach(() => {
    service = new UsersService()

    const fnNames = Object.keys(repo) as (keyof typeof repo)[]
    fnNames.forEach(fnName => {
      jest.spyOn(repo, fnName)
    })
  })

  describe('getAll', () => {
    it('should return all the users', () => {
      expect(service.getAll()).toMatchSnapshot()
      expect(repo.getAllUsers).toHaveBeenCalledWith()
      expect(repo.getAllUsers).toHaveBeenCalledTimes(1)
    })
  })

  describe('search', () => {
    it('should return some of the users', () => {
      expect(service.search('stark')).toMatchSnapshot()
      expect(repo.getUsersByQuery).toHaveBeenCalledWith('stark')
      expect(repo.getUsersByQuery).toHaveBeenCalledTimes(1)
    })
  })

  describe('create', () => {
    let user: IUser

    beforeEach(() => {
      user = service.create('the dude', 'admin')
    })

    it('should create the new user', () => {
      expect(user).toMatchSnapshot()
      expect(repo.addUser).toHaveBeenCalledWith(user)
      expect(repo.addUser).toHaveBeenCalledTimes(1)
      expect(service.search('dude')).toMatchSnapshot()
    })
  })

  describe('update', () => {
    beforeEach(() => {
      service.update({
        avatar: 'lebowski',
        id: 6,
        name: 'Jeff',
        username: 'admin-cant-change-this',
      })
    })

    it('should update the user', () => {
      expect(service.search('admin')).toMatchSnapshot()
    })
  })
})
