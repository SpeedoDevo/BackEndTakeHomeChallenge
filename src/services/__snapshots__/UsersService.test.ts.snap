// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersService create should create the new user 1`] = `
User {
  "avatar": undefined,
  "id": 6,
  "name": "the dude",
  "username": "admin",
}
`;

exports[`UsersService create should create the new user 2`] = `
Array [
  User {
    "avatar": undefined,
    "id": 6,
    "name": "the dude",
    "username": "admin",
  },
]
`;

exports[`UsersService getAll should return all the users 1`] = `
Array [
  User {
    "avatar": undefined,
    "id": 1,
    "name": "Daenerys Targaryen",
    "username": "thequeenbee",
  },
  User {
    "avatar": undefined,
    "id": 2,
    "name": "Jon Snow",
    "username": "forthebrotherhood",
  },
  User {
    "avatar": undefined,
    "id": 3,
    "name": "Sansa Stark",
    "username": "joffery.fan",
  },
  User {
    "avatar": undefined,
    "id": 4,
    "name": "Ned Stark",
    "username": "honouraboveall",
  },
  User {
    "avatar": undefined,
    "id": 5,
    "name": "Cersei Lannister",
    "username": "bestmuminwesteros",
  },
]
`;

exports[`UsersService search should return some of the users 1`] = `
Array [
  User {
    "avatar": undefined,
    "id": 3,
    "name": "Sansa Stark",
    "username": "joffery.fan",
  },
  User {
    "avatar": undefined,
    "id": 4,
    "name": "Ned Stark",
    "username": "honouraboveall",
  },
]
`;

exports[`UsersService update should update the user 1`] = `
Array [
  User {
    "avatar": "lebowski",
    "id": 6,
    "name": "Jeff",
    "username": "admin",
  },
]
`;
